### PR TITLE
Create Alliance Endpoint

### PR DESCRIPTION
**Pull request by thenathanyang**
The `/createAlliance` endpoint (`createAlliance(string[] allies, string[] targets)`) creates a new alliance and adds the alliance ID to users in the alliance.

According to the API Layout document, this endpoint will be called by `joinAlliance()` if an alliance does not exist yet. Here's more info on `joinAlliance()`:

> `joinAlliance(string allianceID=None)`
> This function should (scan/connect?) to others in the area using WiFi Direct. It will call createAlliance(allies, targets) if the alliance does not exist yet. This function will also check to see if any player has any other player within the new alliance as a target - if so, the joinAlliance request will fail.

### CODE REVIEW

File: BackEnd/routes.js [hunk 1]
```diff
@@ -186,6 +188,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+
+router.post('/createAlliance', (req, res) => {
+	var jsonRequestBody = req.body;
+
+	var allianceFields = {};
+	allianceFields.allies = jsonRequestBody.allies;
+	allianceFields.targets = jsonRequestBody.targets;
+
+	var alliance = new Alliance(allianceFields);
+	var allianceId = alliance.id;
```
Comment (prathyushk) [inline BackEnd/routes.js]: isn't the field supposed to be _id?

File: BackEnd/routes.js [hunk 2]
```diff
@@ -186,6 +188,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+	var allies = JSON.parse(jsonRequestBody.allies);
+
+	alliance.save(function(err) {
+		if (err) {
+			console.log("Failed to create alliance - Error: " + err.message);
+			return;
+		}
```
Comment (prathyushk) [inline BackEnd/routes.js]: you might want to put the update in here just ti make sure we don't add an invalid alliance to the players

File: BackEnd/routes.js [hunk 3]
```diff
@@ -186,6 +188,43 @@ router.get('/getTargetLocation', (req, res) => {
 	    res.send(obj.location);
 	}
     });
+
+router.post('/createAlliance', (req, res) => {
+	var jsonRequestBody = req.body;
+
+	var allianceFields = {};
+	allianceFields.allies = jsonRequestBody.allies;
+	allianceFields.targets = jsonRequestBody.targets;
```
Comment (prathyushk) [inline BackEnd/routes.js]: might be better to generate this from the list of allies than get it from the request. we also need to verify that there is no player that appears in both allies and targets

File: BackEnd/routes.js [hunk 4]
```diff
@@ -209,6 +211,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+
+	var alliance = new Alliance(allianceFields);
+	var allianceId = alliance.id;
+	var allies = JSON.parse(jsonRequestBody.allies);
+
+	alliance.save(function(err) {
+		if (err) {
+			console.log("Failed to create alliance - Error: " + err.message);
+			return;
```
Comment (tanya-al) [inline BackEnd/routes.js]: When the other post requests fail, we usually return something like
res.status(400).json({
			error: 'Must have both login code and orgName specified',
		});
Can you edit your code to have a similar error message for consistency?

File: BackEnd/routes.js [hunk 5]
```diff
@@ -209,6 +211,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+		}, {
+			// Update
+			alliance : allianceId,
+		}, {
+			// Options - new: return the modified document rather than the original
+			new : true,
+		}, function(err, doc) {
+			if (err) {
+				console.log("Failed to add users to a new alliance - Error: " + err.message);
+				return;
```
Comment (tanya-al) [inline BackEnd/routes.js]: Same as above

File: BackEnd/routes.js [hunk 6]
```diff
@@ -209,6 +211,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+	alliance.save(function(err) {
+		if (err) {
+			console.log("Failed to create alliance - Error: " + err.message);
+			return;
+		}
+	});
+
+	// Update users with their alliance ID
+	for (int i = 0; i < allies.length; i++) {
+		Player.findOneAndUpdate({
```
Comment (tanya-al) [inline BackEnd/routes.js]: Can we change the target type in Player schema to an array to add the targets of each alliance member to each other member's target list? That would vastly simplify killTarget...

**Review by tanya-al — CHANGES_REQUESTED** [review]

File: BackEnd/routes.js [hunk 7]
```diff
@@ -209,6 +211,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+
+	var allianceFields = {};
+	allianceFields.allies = jsonRequestBody.allies;
+	allianceFields.targets = jsonRequestBody.targets;
+
+	var alliance = new Alliance(allianceFields);
+	var allianceId = alliance.id;
+	var allies = JSON.parse(jsonRequestBody.allies);
+
+	alliance.save(function(err) {
```
Comment (rohan-varma) [inline BackEnd/routes.js]: `alliance.save(err => ..)`

File: BackEnd/routes.js [hunk 8]
```diff
@@ -209,6 +211,43 @@ router.get('/getTargetLocation', (req, res) => {
...
+		}, {
+			// Options - new: return the modified document rather than the original
+			new : true,
+		}, function(err, doc) {
+			if (err) {
+				console.log("Failed to add users to a new alliance - Error: " + err.message);
+				return;
+			}
+		});
+	}
```
Comment (rohan-varma) [inline BackEnd/routes.js]: Also end the req/res cycle if it reaches here (indicating success I assume)

File: BackEnd/routes.js [hunk 9]
```diff
@@ -236,6 +238,54 @@ router.get('/getTargetLocation', (req, res) => {
...
+			console.log("Failed to create alliance - Error: " + err.message);
+			res.status(400).json({
+				error: "Failed to create alliance - Error: " + err.message,
+			});
+			return;
+		}
+	});
+
+	// Update users with their alliance ID
+	for (int i = 0; i < allies.length; i++) {
```
Comment (rohan-varma) [inline BackEnd/routes.js]: use `let`, `int` will cause this to crash

Comment (thenathanyang) [reply BackEnd/routes.js]: Oops my bad lol

**Review by tanya-al — APPROVED** [review]

**Review by prathyushk — APPROVED** [review]

**Comment by thenathanyang** [comment]
New PR: [https://github.com/rohan-varma/CS-M117-Project/pull/23](url)

**Pull request closed without merging**